course/src/Isometry.cc

```diff
@@ -96,21 +98,152 @@ const Vector3 Vector3::kZero(0.0, 0.0, 0.0);
 
 // Matrix3
 
-const Matrix3 Matrix3::kIdentity({1.0, 0.0, 0.0}, {0.0, 1.0, 0.0}, {0.0, 0.0, 1.0});
-const Matrix3 Matrix3::kZero({0.0, 0.0, 0.0}, {0.0, 0.0, 0.0}, {0.0, 0.0, 0.0});
+Matrix3::Matrix3(const std::initializer_list<double>& row_1, const std::initializer_list<double>& row_2,
+                 const std::initializer_list<double>& row_3) {
+    mat_.resize(3);
+    if ((row_1.size() < 3) || (row_2.size() < 3) || (row_3.size() < 3)) {
+        throw "Size of the matrix isn't correct";
+    }
+    auto it = row_1.begin();
+    mat_[0][0] = *it++;
+    mat_[0][1] = *it++;
+    mat_[0][2] = *it++;
+
+    it = row_2.begin();
+    mat_[1][0] = *it++;
+    mat_[1][1] = *it++;
+    mat_[1][2] = *it++;
+
+    it = row_3.begin();
+    mat_[2][0] = *it++;
+    mat_[2][1] = *it++;
+    mat_[2][2] = *it++;
+}
+Matrix3::Matrix3(const double m00, const double m01, const double m02, const double m10, const double m11,
+                 const double m12, const double m20, const double m21, const double m22)
+    : Matrix3{{m00, m01, m02}, {m10, m11, m12}, {m20, m21, m22}} {}
+
+Matrix3::Matrix3() : Matrix3{{0.0, 0.0, 0.0}, {0.0, 0.0, 0.0}, {0.0, 0.0, 0.0}} {}
+
+Matrix3::Matrix3(const Matrix3& mat)
+    : Matrix3{mat[0][0], mat[0][1], mat[0][2], mat[1][0], mat[1][1], mat[1][2], mat[2][0], mat[2][1], mat[2][2]} {}
+
+Matrix3::Matrix3(Matrix3&& mat)
+    : Matrix3{mat[0][0], mat[0][1], mat[0][2], mat[1][0], mat[1][1], mat[1][2], mat[2][0], mat[2][1], mat[2][2]} {}
+
+Matrix3& Matrix3::operator=(Matrix3&& mat) {
+    mat_ = std::move(mat.mat_);
+    return *this;
+}
+
+Matrix3& Matrix3::operator=(const Matrix3& mat) {
+    // check for self-assignment
+    if (&mat == this) return *this;
+    // reuse storage when possible
+    mat_ = mat.mat_;
+    return *this;
+}
 
 bool Matrix3::operator==(const Matrix3& a) const {
-    auto it = values_.begin();
-    auto a_it = a.values_.begin();
-    for (uint8_t count = 0; count < values_.size(); ++count) {
-        if (!almost_equal(*a_it, *it, resolution)) {
+    for (auto idx = 0; idx < 3; ++idx) {
+        if (mat_[idx] != a[idx]) {
             return false;
         }
-        ++it;
-        ++a_it;
     }
     return true;
 }
 
+bool Matrix3::operator==(const std::initializer_list<double>& rhs) const {
+    if (rhs.size() < 9) {
+        return false;
+    }
+    auto it = rhs.begin();
+
+    for (auto idx = 0; idx < 3; ++idx) {
+        if (mat_[idx] != std::initializer_list<double>({*it++, *it++, *it++})) {
```

Comment:
Look at this comment: https://github.com/LucasJSch/cppl1_q12020/pull/2#discussion_r381283578